### PR TITLE
Mary Ampers

### PR DESCRIPTION
**Pull request by mmlamkin**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | The Edamam site had good documentation about consuming their API and what was needed when using it. I used postman to test calls 
Describe your API Wrapper.  How did you decide on the methods you created? | My api wrapper has two methods. One makes a call through HTTParty to get a list of recipes including the word searched for. It makes each recipe into a Recipe class object, however I'm not sure it was totally necessary to get all of that information for the index page of the site. The other method makes a call to the API for a specific recipe for the 'show' page of the site. It uses the recipe's encoded URI and creates a Recipe object from the information from the API
Describe an edge case or failure case test you wrote for your API Wrapper. | I tested for the case in which a query doesn't bring back any recipes. This should just return an empty array rather than throw an error, which could happen if the method tried to create a Recipe object out of a return nil.
Explain how VCR aids in testing an API. | Since API's usually only allow a certain number of calls VCR records a call from the API and allows you to reuse that information for multiple tests rather than making more calls
What is the Heroku URL of your deployed application? | https://mary-muncher.herokuapp.com
Provide a link to the Trello board you used | https://trello.com/b/wUebYDpB/muncher

### CODE REVIEW

File: test/controllers/muncher_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        recipes = EdamamApiWrapper.list_recipes("chicken")
+        must_respond_with :ok
+      end
+      must_respond_with :success
+    end
+
+    it "should redirect to root path with nil or empty string" do
+      VCR.use_cassette("recipe") do
+        get muncher_index_path
+        recipes = EdamamApiWrapper.list_recipes("")
```
Comment (CheezItMan) [inline test/controllers/muncher_controller_test.rb:27]: Again this test is operating on the wrapper rather than the controller.  

Your test should look like:
```ruby
    it "should redirect to root path with nil or empty string" do
      VCR.use_cassette("recipe") do
        get muncher_index_path, params: {search: ""}
        must_respond_with :redirect
        must_redirect_to root_path
      end
    end
```

File: test/controllers/muncher_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    it "should succeed with a valid uri" do
+      VCR.use_cassette("recipe") do
+        get muncher_path("7bf4a371c6884d809682a72808da7dc2")
+      end
+      must_respond_with :success
+    end
+
+    it "should redirect to the root path with a bogus uri" do
+      VCR.use_cassette("recipe") do
```
Comment (CheezItMan) [inline test/controllers/muncher_controller_test.rb:45]: This test is failing see my notes on the Recipe class and wrapper

File: lib/recipe.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+class Recipe
+  attr_accessor :name, :uri, :ingredients, :image, :dietary_info,:url, :source, :id
+
+  def initialize(name, uri, ingredients, image, dietary_info, url, source)
+    @name = name
```
Comment (CheezItMan) [inline lib/recipe.rb:5]: You need to verify the arguments to this method otherwise it will fail to split and crash the site.

File: lib/edamam_api_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      encode = URI.encode("#{URL}?r=#{URL2}#{id}&app_id=#{APP_ID}&app_key=#{APP_KEY}")
+
+      response = HTTParty.get(encode)
+
+      if response.empty?
+        
+        return nil
+
+      else
+
```
Comment (CheezItMan) [inline lib/edamam_api_wrapper.rb:36]: You need to verify that the response has the keys below, for invalid ids this will crash the app.

File: app/views/muncher/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+</header>
+
+<h4 class='results'>results for "<%= params[:search] %>"</h4>
+<section class='recipes'>
+
+<% if @recipes %>
+    <ul>
+      <% @recipes.each do |recipe| %>
+        <li>
+          <img src="<%= recipe.image %>" alt="recipe photo">
```
Comment (CheezItMan) [inline app/views/muncher/index.html.erb:19]: Making this image a link would improve the user experience.

File: test/controllers/muncher_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+    it "should redirect to root path with nil or empty string" do
+      VCR.use_cassette("recipe") do
+        get muncher_index_path
+        recipes = EdamamApiWrapper.list_recipes("")
+        must_respond_with :redirect
+        must_redirect_to root_path
+      end
+
+    end
```
Comment (CheezItMan) [inline test/controllers/muncher_controller_test.rb:32]: You should also test where the search term is missing.

File: test/lib/recipe_test.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    proc {
+      Recipe.new
+    }.must_raise ArgumentError
+
+    proc {
+      Recipe.new("Name")
+    }.must_raise ArgumentError
+  end
+
+  it "must initialize parameters correctly" do
```
Comment (CheezItMan) [inline test/lib/recipe_test.rb:15]: You should also test an ID that you can't split

**Comment by CheezItMan** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | More commits would be better, your commit messages are good
Comprehension questions | Check, 
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | Your routes are **not** restful.  A more restful set of routes would be `/recipes` for the list of recipes and `/recipes/:id` for the show action.
Semantic HTML | Check
Errors are reported to the user | Check
API Wrapper to handle the API requests | . Check 
Controller testing | Check, but you have some troubled tests here, see my notes.  You are mixing wrapper and controller tests here.
Lib testing | Check
Search Functionality | Check
List Functionality |  Check
Show individual item functionality (link to original recipe opens in new tab) | Check
**Styling** | 
Responsive layout | Both the search results and show pages are responsive.  
List View shows 10 items at a time/pagination | Check
The app is styled to create an attractive user interface | Fairly simple, but attractive style
**API Features** | 
The App attributes Edaman | Check
The VCR casettes do not contain the API key | 
**External Resources** | 
Link to Trello Board | Check
Link to deployed app on Heroku | Check
**Overall** |  You hit most of the learning goals.  You did have some serious problems with your testing.  Check my notes in your code and let me know if you have questions.
